src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -657,6 +714,147 @@ protected virtual void NotifyListeners(ProjectChangeEventArgs e)
         }
     }
 
+    internal override void UpdateProject(
+        string filePath,
+        RazorConfiguration configuration,
+        ProjectWorkspaceState projectWorkspaceState,
+        string rootNamespace,
+        Func<IProjectSnapshot, ImmutableArray<IUpdateProjectAction>> calculate)
+    {
+        if (projectWorkspaceState is null)
+        {
+            throw new ArgumentNullException(nameof(projectWorkspaceState));
+        }
+
+        // Get an upgradeableLock, which will keep a read lock while we compute the changes
+        // and then get a write lock to actually apply them. Only one upgradeable lock
+        // can be held at any given time. Write lock must be retrieved on the same
+        // thread that the lock was acquired
+        using var upgradeableLock = _locksFactory.EnterUpgradeAbleReadLock();
+
+        var project = GetLoadedProject(filePath);
+        if (project is not ProjectSnapshot projectSnapshot)
+        {
+            return;
+        }
+
+        var originalHostProject = projectSnapshot.HostProject;
+        var changes = calculate(project);
```

Comment:
My hope is basically everything here is just and `ApplyUpdate(UpdateAction action)` 'ApplyUpdates(params UpdateAction[] actions)` at some point. It indicates what mutations are possible, makes sure information is calculated up front, etc. For now it's not quite that simple to migrate :/

Reply:
That seems to match perfectly with `ImmutableInterlocked.Update`, but I don't expect you to predict whether it will be possible, or even desirable, to move to that model instead. Just an observation :)